activerecord/lib/active_record/normalization.rb

```diff
@@ -0,0 +1,151 @@
+      # and the normalized values will be persisted to the database.
+      # Normalizations are also applied to matching keyword arguments of finder
+      # methods. This allows a record to be created and later queried using an
+      # unnormalized value.
+      #
+      # However, to prevent confusion, normalizations are not applied when
+      # attributes are fetched from the database. This means that if a record
+      # was persisted before a normalization was declared, the relevant
+      # attributes will not be normalized until either they are assigned new
+      # values, or they are explicitly migrated via +normalize_attribute+.
+      #
+      # Because normalizations may be applied multiple times, they should be
+      # _idempotent_. In other words, applying a normalization more than once
+      # should have the same result as applying it only once.
+      #
+      # Normalizations, by default, are not applied to +nil+ values. This
+      # behavior can be changed with the +:including_nil_values+ option.
+      #
+      # ==== Options
+      #
+      # * <tt>:with</tt> - The normalization to apply. May be specified as
+      #   either a callable or a Symbol which will be converted to a Proc.
+      # * <tt>:including_nil_values</tt> - Whether to apply normalizations to
+      #   +nil+ values. Defaults to +false+.
+      #
+      # ==== Examples
+      #
+      #   class User < ActiveRecord::Base
+      #     normalizes :email, with: -> email { email.strip.downcase }
+      #   end
+      #
+      #   user = User.create(email: " CRUISE-CONTROL@EXAMPLE.COM\n")
+      #   user.email                  # => "cruise-control@example.com"
+      #
+      #   user = User.find_by(email: "\tCRUISE-CONTROL@EXAMPLE.COM ")
+      #   user.email                  # => "cruise-control@example.com"
+      #   user.email_before_type_cast # => "cruise-control@example.com"
+      #
+      #   User.where(email: "\tCRUISE-CONTROL@EXAMPLE.COM ").first       # => user
+      #   User.where("email = ?", "\tCRUISE-CONTROL@EXAMPLE.COM ").first # => nil
+      def normalizes(*names, including_nil_values: false, with:)
+        with = with.to_proc if with.is_a?(Symbol)
+
+        names.each do |name|
+          attribute(name) do |cast_type|
+            NormalizedValueType.new(cast_type: cast_type, normalizer: with, normalize_nil: including_nil_values)
+          end
+        end
+
+        self.normalized_attributes += names.map(&:to_sym)
+      end
+
+      # Normalizes a given +value+ using normalizations declared for +name+.
+      #
+      # ==== Examples
+      #
+      #   class User < ActiveRecord::Base
+      #     normalizes :email, with: -> email { email.strip.downcase }
+      #   end
+      #
+      #   User.normalize(:email, " CRUISE-CONTROL@EXAMPLE.COM\n")
+      #   # => "cruise-control@example.com"
+      def normalize(name, value)
+        type_for_attribute(name).cast(value)
+      end
+    end
```

Comment:
Normalization is done via type casting (and normalizers expect a type-casted value).  `ClassMethods#normalize` knows this, so it calls `cast`.  The outside world does not know this, and does not need to know, so they can call `ClassMethods#normalize`.

I should also point out that we cannot call `type_for_attribute(name).normalize(value)` because there might be another type decoration / wrapper stacked on top of `NormalizedValueType`.  There might even be two `NormalizedValueType`s stacked on top of each other, and calling `NormalizedValueType#normalize` would apply only one normalization, not both.